### PR TITLE
SOLR-16271: remove wildcard imports

### PR DESCRIPTION
**Pull request by epugh**
https://issues.apache.org/jira/browse/SOLR-16271

# Description

Remove wildcard imports from the build.   In later issue, will add a spotless based check.

# Solution

made the changes.

# Tests

manually run tests.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by epugh** [comment]
Just pushed up the spotless version.   We don't get the nice report, since it just stops on the first one with an exception:

```
> Task :solr:core:spotlessJava FAILED
Step 'Refuse wildcard imports' found problem in 'src/test/org/apache/solr/handler/TestReplicationHandler.java':
Do not use wildcard imports.  'spotlessApply' cannot resolve this issue.
java.lang.AssertionError: Do not use wildcard imports.  'spotlessApply' cannot resolve this issue.
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:72)
        at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
        at spotless_9enkbgsd2kcvij6tpbbr3dzhe$_run_closure1$_closure2$_closure8$_closure10$_closure11.doCall(/Users/epugh/Documents/projects/solr-epugh-2/gradle/validation/spotless.gradle:66)

```

**Comment by epugh** [comment]
@dsmiley before I go futhur, what do you think of the way I fixed the imports?   Does this pattern for both regular and statics make sense?

**Comment by dsmiley** [comment]
No wildcards -- not because of me or you or whatever IntelliJ or other IDE ships by default, but because we are trying to adhere to a particular style guide: https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

**Comment by epugh** [comment]
awesome, so, is everything in this style guide "solr community accepted"?   As in, can I reference this to understand what our style is?   We should have a FAQ entry about that for the next person ;-)

**Comment by uschindler** [comment]
Lucene agreed to the Spotless code formatting with the Google Java. In Lucene's precommit will fail, if you have code not formatted like that. In general it's plain easy: run "./gradlew tidy" before commit. No need to take care anything. If your IDE has slightly different style it does not matter if you run the command before commit. 👍

**Comment by uschindler** [comment]
I don't think this PR is needed. I'd just go with spotless.

**Comment by epugh** [comment]
I believe that we have to manually deal with the import statements, that spotless can alert, but can't fix it....?

**Comment by uschindler** [comment]
From my knowledge, spotless can deal with imports. It uses javac to scan the code, so it understands how to expand wildcards.

**Comment by dsmiley** [comment]
Our repo already documents adherence to "Google Java Format" -- https://github.com/apache/solr/blob/main/help/formatting.txt but perhaps we should restate this in more places.

Spotless supports multiple formats, including "Google Java Style Guide", [but it has limitations].(https://github.com/diffplug/spotless/issues/649).

Also CC @dweiss as you may be interested in this.

**Comment by epugh** [comment]
> From my knowledge, spotless can deal with imports. It uses javac to scan the code, so it understands how to expand wildcards.

From my investigation, spotless can reorder imports, but not expand them....   Any links/docs on how to enable it?

**Comment by dsmiley** [comment]
Just a quick tip: IntelliJ can do all the work here across the project; no need to open any source file at all.  Just select the root of the project in the navigation tree and choose "Optimize Imports" in the Code menu.  It might be influenced by having the "Google-java-format" plugin enabled.

**Comment by dsmiley** [comment]
To keep the build changes clearly visible from a massive commit that makes us adhere, I propose doing this in two PRs; first actually optimizes imports via the IDE (a _different_ PR), and the second (this PR) then enforces the rule.

**Comment by epugh** [comment]
> Just a quick tip: IntelliJ can do all the work here across the project; no need to open any source file at all. Just select the root of the project in the navigation tree and choose "Optimize Imports" in the Code menu. It might be influenced by having the "Google-java-format" plugin enabled.

Trying it now!   Still learning my way around intellij, thanks for sharing!

**Comment by epugh** [comment]
@dsmiley the organize imports work great, however intellij had different order then spotless, so re-ran spotless to restore orginal order ,and now I am about to push up the explicit import changes.    

Will move the enforcement code to another issue.

**Comment by madrob** [comment]
Just a note that some imports exist because of javadoc references and I think automated tools sometimes has issues in the past with removing them too aggressively. Worth a check after changes instead of blindly committing, and if you see it hopefully this is enough warning that you won't be surprised :)

**Comment by epugh** [comment]
@madrob I did notice the ` // jdoc` got removed....    I never did figure out what they did!

**Comment by epugh** [comment]
I ran `gw clean` and then `gw assemble` and the ref guide seemed okay.   Linked to the solr javadocs....    The solrj pulling in code in the ref guide seemed good as well.

Other suggestions on what to verify?

File: solr/core/src/java/org/apache/solr/parser/QueryParser.java [hunk 1]
```diff
@@ -2,22 +2,21 @@
...
 
-import org.apache.lucene.analysis.Analyzer;
-import org.apache.lucene.search.BooleanClause;
-import org.apache.lucene.search.Query;
-import org.apache.solr.search.SyntaxError;
-import org.apache.solr.search.QParser;
-
-import org.apache.lucene.queryparser.charstream.CharStream;
-import org.apache.lucene.queryparser.charstream.FastCharStream;
-
```
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/parser/QueryParser.java:20]: minor: technically not wildcard import removal

**Comment by epugh** [comment]
> > I did notice the // jdoc got removed.... I never did figure out what they did!
> 
> It looks like the comment was the only thing removed, not the import itself. If we can leave the comment somehow then I'd try to keep it, otherwise somebody in the future is inevitably going to try to flag those as unused imports and then stub their toe on it (that person is going to be me, I am the one that will stub his toe)

are you saying the text ` // jdoc` is literally the comment?   COuld we expand it to be something like ` // import to support Javadocs` or something like that?   ` // jdoc` just didn't really mean anything to me!

**Comment by epugh** [comment]
I imagined ` // jdoc` to not be a comment, but to be some sort of crazy annotation or maybe a reminder to someone to come back and "jdoc" that line (whatever that meant!)...   So a more verbose comment would make it clearer...

**Comment by epugh** [comment]
Here is what I found for the text `jdoc` in the source:
image

**Comment by epugh** [comment]
Okay, I think I am starting to understand...   For the example of `SwitchQParserPlugin.java`, we reference `SearchHandler` in the javadocs, and so the import is needed.  If I remove the import, then IntelliJ does flag it that it is needed....    

So, I guess, if we only have those three files that follow this pattern, then we can leave the comment in, but maybe change it from `// jdoc` to `// support javadoc linking` ???    

Or, maybe since intellij flags the missing import if you don't have it, then we don't need the comment at all?

**Review by cpoerschke — APPROVED** [review]
LGTM but also okay with `jdoc` comments being retained or changed to something similar.

**Review by uschindler — APPROVED** [review]

**Pull request merged**